drivers: eeprom: add API for EEPROM devices

henrikbrixandersen opened the pull request:
Add API for accessing Electrically Erasable Programmable Read-Only Memory (EEPROM) devices.

EEPROMs have an erase block size of 1 byte, a long lifetime, and allows overwriting data on byte-by-byte access.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: eeprom: add API for EEPROM devices

Add API for accessing Electrically Erasable Programmable Read-Only
Memory (EEPROM) devices.

EEPROMs have an erase block size of 1 byte, a long lifetime, and allows
overwriting data on byte-by-byte access.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
The idea behind this is a simple, byte-oriented API for EEPROM devices.

I have been implementing drivers for generic SPI EEPROMs (AT25xxx compatible) and I2C EEPROMs (AT24xxx compatible) under the Zephyr flash API, but it just doesn't feel right.

The EEPROM drivers need to implement the full flash API (What to do on erase? NO-OP? Write 0xFF to "erased" bytes thus affecting lifetime negatively? Something else?) and this type of devices generally aren't fit as backends for the various subsystems in Zephyr currently supporting the flash API as backend (e.g. the settings subsystem) due to assumptions about page layout and erase.

I would like to hear any comments and/or suggestions on how to approach integrating EEPROM devices in as first-class devices in Zephyr.

Laczen commented:
@henrikbrixandersen, nice idea to add EEPROM devices support to zephyr. All backend systems for zephyr have been written to use flash devices for storage and they will probably not be an ideal fit for EEPROM devices. But even a non ideal fit would allow these devices to be used:

a. The flash storage solutions try to do everything to avoid erasing pages, so even writing 0xff (I would do this) as a erase operation would not influence lifetime a lot.
b. The flash page size can be selected smaller (e.g. 512 bytes).

Since page erases and page sizes are not EEPROM properties I would separate these properties in a "eeprom_flash" library, so not directly in the EEPROM library. The "glue" between eeprom and flash could be similar to the flash simulator (of course simplified).

This proposed solution would not take full advantage of the EEPROM devices, to do that a storage method for EEPROM devices needs to be created.

nvlsianpu commented:
@henrikbrixandersen 
Maybe worth to add a flash driver which will be a wrapper onto eeprom API.
Storage system dedicated to eeprom should be designed slight different than for flash.

nvlsianpu commented:
>EEPROMs typically have an erase block size of 1 byte and a long

have you any example of an EEPROM with different garrulity - If so, need to add API for fetch this.

pabigot reviewed (commented):
It makes sense to do something simpler than the Zephyr flash API, and this seems like a reasonable approach.

We do need devicetree bindings for the EEPROMs.  There's an existing `atmel,at24` binding in Zephyr under mtd that supports a hand-coded sample tests/drivers/i2c/i2c_slave_api for I2C eeprom.  It's ancient and should be replaced by one closer to [Linux's](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/eeprom/at24.txt), with a companion [atmel,25](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/eeprom/at25.txt) binding added.

Is it true that all the eeprom variation points can be parameterized in devicetree, so there would be a single implementation of this driver?

It should support multiple instances (though that may be difficult until Zephyr's devicetree becomes more mature), as well as support both I2C and SPI as the back-end.  There are some sensor drivers that support both busses that could be used as an example, or we could consider take advantage of the opportunity and provide a generic block back-end that's independent of bus.

aunsbjerg commented:
Looks good!

Some EEPROM chip's have dedicated erase commands. Maybe the API could support this by having something like `erase(struct *dev, off_t offset, size_t size)` method as well?

henrikbrixandersen commented:
> Looks good!

Thank you for reviewing :+1: 

> Some EEPROM chip's have dedicated erase commands. Maybe the API could support this by having something like `erase(struct *dev, off_t offset, size_t size)` method as well?

Wouldn't they fit under our existing flash API, then?

henrikbrixandersen commented:
> It makes sense to do something simpler than the Zephyr flash API, and this seems like a reasonable approach.

Thank you for reviewing :+1: 
 
> We do need devicetree bindings for the EEPROMs. There's an existing `atmel,at24` binding in Zephyr under mtd that supports a hand-coded sample tests/drivers/i2c/i2c_slave_api for I2C eeprom. It's ancient and should be replaced by one closer to [Linux's](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/eeprom/at24.txt), with a companion [atmel,25](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/eeprom/at25.txt) binding added.

Yes, I have those bindings here. I will add them for review.

> Is it true that all the eeprom variation points can be parameterized in devicetree, so there would be a single implementation of this driver?

Well, yes. But I would prefer having a dedicated `eeprom-at24.c` and a dedicated `eeprom-at25.c` driver, since the underlying bus access is what makes up most of the driver. Putting both I2C and SPI support into the same driver does not give any real benefit, I think.

henrikbrixandersen commented:
> have you any example of an EEPROM with different garrulity - If so, need to add API for fetch this.

No, I haven't come across any. I will reword the sentence to reflect that.

henrikbrixandersen commented:
Regarding a flash API adapter layer: Good idea, if anybody has a use for that. Another option could be to implement EEPROM-specific backends for the various subsystems that could use EEPROMs as backends when the need arises.

I would like to keep the first implementation as simple as possible and focus on the hardware-near API.

Laczen commented:
@henrikbrixandersen, yes I think you should keep the first implementation as simple as possible and focus on the hardware-near API. An EEPROM has different properties than a flash device and the EEPROM API should reflect this -> so I think your proposal is the correct way.

When you would like to use it for (settings) storage you could define a flash_eeprom API layer, that is just a wrapper to your read/write methods but also defines a virtual page size (e.g 512 bytes) to define a "flash_erase()" and also to allow the storage backends to get a page size. This layer could then be used as a "flash" interface to store all settings in EEPROM without any modifications.

nvlsianpu commented:
> Some EEPROM chip's have dedicated erase commands. 

@aunsbjerg Can you point such a part? Does it men that for such a device need to erase an area befor write to it, or it is just a helper feature for increase write speed.

nvlsianpu commented:
> I would like to keep the first implementation as simple as possible and focus on the hardware-near API.

@henrikbrixandersen Right. My comment was suited for pointing direction which can satisfy flash API client - this can be added in subsequent PR by anyone who needs such a solution.

aunsbjerg commented:
> > Some EEPROM chip's have dedicated erase commands.
> 
> @aunsbjerg Can you point such a part? Does it men that for such a device need to erase an area before write to it, or it is just a helper feature for increase write speed.

The Microchip 25LC1024 has chip/page/sector erase commands. They are only helper features for quicker erase operation. The same result could be achieved by manually erasing using write operations.

I don't consider an `erase()` method in the API to be crucial at all, but for those EEPROM devices that do support it, it would be nice to have without having to "hide" an EEPROM behind the flash interface.

henrikbrixandersen commented:
@aunsbjerg 
> I don't consider an `erase()` method in the API to be crucial at all, but for those EEPROM devices that do support it, it would be nice to have without having to "hide" an EEPROM behind the flash interface.

That leads to another questions: If we have to have an `erase()` API method in the EEPROM API (which most of the EEPROM device drivers would likely not implement or just provide a dummy method for), would we be better off just "hiding" all EEPROM devices behind the flash API?

Personally, I still believe a dedicated, simple EEPROM API is the way to go. For devices which have the concept of explicit erase, I would prefer to use the flash API.

pabigot commented:
> The Microchip 25LC1024 has chip/page/sector erase commands. They are only helper features for quicker erase operation. The same result could be achieved by manually erasing using write operations.

That chip implements the full JEDEC API so is already supported by the flash spi_nor driver.  My understanding is that this API targets much more limited chips that don't have such erase operations.

I tend to agree with @henrikbrixandersen that there is no need for an erase function in this API.  There may be a value in a library function that implements "write N copies of sequence S with length L starting at offset O" which can support erase protocols of varying types (write 0xFF, write patterns, etc.).

nvlsianpu commented:
@aunsbjerg Aggre - with erase API it will be almost flash API. I'm in line with that this should be covered by flash API - eeprom api should be simpler.

alexanderwachter commented on include/drivers/eeprom.h:
```diff
@@ -0,0 +1,137 @@
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write(struct device *dev, off_t offset, const void *data,
```
Why do you need a syscall here? As I understand the API, you will only use the i2c or SPI APIs, which are syscalls themselves.

henrikbrixandersen replied:
The two drivers, I have in progress, only use the I2C and/or SPI APIs, correct. But other EEPROM driver can be added at a later point (e.g. the NXP Kinetis series has the ability to expose part of its flash as EEPROM).

Laczen reviewed (approved):
@henrikbrixandersen do you have any plans to provide a eeprom simulator ? If not I might take a look at it, should be very similar to flash_simulator and can help a lot during development and testing of systems that are using eeprom.

alexanderwachter commented on include/drivers/eeprom.h:
```diff
@@ -0,0 +1,142 @@
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->size;
```
api->size() ?

nvlsianpu replied:
const size_t size;

alexanderwachter replied:
Sorry, I overlooked that. I didn't expect something like a constant (not a function ptr) in a struct called API. But, the flash API seems to do this in the same way.

nvlsianpu commented on include/drivers/eeprom.h:
```diff
@@ -0,0 +1,142 @@
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write_protection_set(struct device *dev, bool enable);
```
Doesn't it make sense to make protection mechanism automatic? User know what is doing while call write API - this driver, so is no trivial like accidentally writing a register in bare-metel code.

nvlsianpu replied:
I think this API can be removed.

Laczen replied:
I think this is similar to the flash idea, by separating it from the write you can have a write_protection_set, many writes, and then a write_protection_set. I think it is good to have this. If the hardware doesn't need this, it can always be a nop.

nvlsianpu replied:
@pabigot Your opinion?

pabigot replied:
While I normally like explicit control of all features, the Linux at24 and at25 drivers manage WREN/WPn within the write function.

Unless there's a functional reason to require external confirmation that the write is intentional, or a significant performance impact for write-enable on each write, I agree that this function *can* be removed and protection-management done within the write function.

Whether it *should* be removed I'll leave to @henrikbrixandersen.

henrikbrixandersen replied:
@nvlsianpu, @Laczen, @pabigot: Thank you for your input on this. I have removed the write protection method from the proposed API.

nvlsianpu commented:
> do you have any plans to provide a eeprom simulator ? If not I might take a look at it, should be very similar to flash_simulator and can help a lot during development and testing of systems that are using eeprom.

I'm not sure whether not better to implement a posix driver.

Laczen commented:
> > do you have any plans to provide a eeprom simulator ? If not I might take a look at it, should be very similar to flash_simulator and can help a lot during development and testing of systems that are using eeprom.
> 
> I'm not sure whether not better to implement a posix driver.

I don't know, I am not used to working with native_posix. Does native_posix work on qemu_x86 ? Can I use this also on a real device (e.g.  nrf52) ? Anyhow it would be easy to add a native_posix interface as well.

nvlsianpu commented:
Native posix executes on machine you compile for (so your PC).
One advantage is that you can debug it using GDB on you machine. Second advantage is that it is possible to produce a raw bin file which represent the eeprom - this file can be used to program a real dice or as NV storage in a simulated test with system reboot inside the suite.
I'm sure you know that such driver is already implemented for flash API, so this shouldn't be difficult.

gon1332 commented:
> Native posix executes on machine you compile for (so your PC).
> One advantage is that you can debug it using GDB on you machine. Second advantage is that it is possible to produce a raw bin file which represent the eeprom - this file can be used to program a real dice or as NV storage in a simulated test with system reboot inside the suite.
> I'm sure you know that such driver is already implemented for flash API, so this shouldn't be difficult.

Both cases are really important. Furthermore, the produced bin file can be really useful for manufacturing purposes.

henrikbrixandersen commented:
@Laczen
> @henrikbrixandersen do you have any plans to provide a eeprom simulator ? If not I might take a look at it, should be very similar to flash_simulator and can help a lot during development and testing of systems that are using eeprom.

I haven't looked into writing an EEPROM simulator, so you're most welcome to look into this. I do intend to provide a POSIX EEPROM driver, though, as I intend to use that for unit tests.

Laczen commented:
@henrikbrixandersen, I have a first version of a eeprom_simulator running.

During the process of development I have been thinking about the write protection. It might still be a good idea to keep this: suppose you have a board were there is some security processor limiting access to the eeprom. Then you would disable the write protection by first sending a command to the security processor and afterwards doing the write. In this case a set_protection() would better fit then a integrated set protection in the eeprom write.

pabigot commented:
> [...] suppose you have a board where there is some security processor limiting access to the eeprom. Then you would disable the write protection by first sending a command to the security processor and afterwards doing the write.  In this case a set_protection() would better fit then a integrated set protection in the eeprom write.

I don't see that.  The protection is associated with the security processor, not the EEPROM, which is a subordinate device.  It seems just like power management: you may have to supply power to the EEPROM before you can interact with it.

Laczen commented:
OK, so there seems to be no need to keep track of the state inside the API for security or pm. We can close this and leave API as proposed.

henrikbrixandersen committed:
dts: bindings: eeprom: add EEPROM device tree bindings

Add device tree binding for Atmel AT25 (and compatible) SPI
EEPROMs. Update the curent Atmel AT24 I2C EEPROM binding to match.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
I have updated this PR with initial DTS bindings for `atmel,at24` and `atmel,at25`. I have not yet updated the `i2c_slave_api` test case to match the new bindings.

henrikbrixandersen committed:
tests: drivers: i2c_slave_api: update overlays to match new dts bindings

Update the board overlays used in the i2c_slave_api tests to match the
new device tree bindings for atmel,at24 I2C EEPROMs.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: i2c: eeprom_slave: update driver to match the new dts bindings

Update the I2C slave EEPROM driver to match the new atmel,at24 device
tree binding, where the size of the EEPROM is specified in bytes.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

alexanderwachter commented on drivers/eeprom/Kconfig:
```diff
@@ -0,0 +1,23 @@
+#
+
+#
+# Flash driver
```
EEPROM?

henrikbrixandersen replied:
@alexanderwachter Fixed, thanks.

gon1332 commented:
Would it be possible for the driver to support multiple instances of EEPROM devices?

gon1332 commented:
@henrikbrixandersen 3 would be enough.

henrikbrixandersen committed:
drivers: eeprom: add EEPROM shell commands

Add shell commands for reading from and writing to an EEPROM device.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: board_shell: enabled EEPROM shell

Enable EEPROM shell commands in the board_shell test application.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
dts: bindings: eeprom: add binding for zephyr native posix EEPROM

Add device tree binding for representing a Zephyr native POSIX EEPROM
device.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: eeprom: add driver for a zephyr native posix EEPROM

Add driver for emulating an EEPROM device using the native POSIX
board. The EEPROM is backed by a binary file in the host file system.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: native_posix: add EEPROM device

Add a 32kB EEPROM device to the Zephyr native POSIX board for testing
purposes.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

aescolar reviewed (dismissed):
I would like to review the EPROM driver for native_posix, but I won't have time until (best case) middle of this week. So I hope you do not mind waiting a little bit (given that you just added that driver yesterday)
So this is not an actual "rejection", just an obvious note for those with merge rights about this.
(If for some reason I do not provide a review soon enough please ping me)

henrikbrixandersen committed:
drivers: eeprom: add driver for Atmel AT24 I2C and AT25 SPI EEPROMs

Add driver for Atmel AT24 (and compatible) I2C along with Atmel AT25
(and compatible) SPI EEPROMs.

Tested with: AT24LC025, AT24C256, AT25AA02E48, and AT25080.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

alexanderwachter commented on drivers/eeprom/eeprom_shell.c:
```diff
@@ -0,0 +1,246 @@
+	int err;
+	int i;
+
+	if (argc < args_no.write) {
```
You can use `SHELL_CMD_ARG` in lines 238-242 and pass the number of mandatory and optional parameters instead of this check here. Same for the read command.

henrikbrixandersen replied:
@alexanderwachter Thanks! I have updated the code to use `SHELL_CMD_ARG()`.

henrikbrixandersen committed:
tests: drivers: build_all: add EEPROM build configuration

Add Atmel AT2x driver to build_all build configuration.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

aescolar commented on drivers/eeprom/eeprom_native_posix.c:
```diff
@@ -0,0 +1,216 @@
+		return -EIO;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
```
Just a silly comment: In native_posix this would be superstitious :),
the memcpy finishes in 0 simulated time and cannot be interrupted.
It does not harm to use a mutex, but it does not help either.

henrikbrixandersen replied:
@aescolar Thanks. I have removed the mutex.

aescolar reviewed (approved):
The native_posix changes seem fine to me.
Thank you very much for adding a driver for it :)

pabigot commented on dts/bindings/mtd/eeprom.yaml:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2019 Vestas Wind Systems A/S
```
I would name this `eeprom-base.yaml`.  Personal preference.

henrikbrixandersen replied:
@pabigot Thank you. I have renamed the file.

pabigot commented on dts/bindings/mtd/eeprom.yaml:
```diff
@@ -0,0 +1,14 @@
+
+include: base.yaml
+
+properties:
```
Is there a reason to not include `size` as a base property?

henrikbrixandersen replied:
The idea was that some EEPROM drivers (e.g. for SoC EEPROMs) might only be able to provide this at run-time. I can move it to `eeprom-base.yaml` if you feel strongly about it :)

pabigot replied:
I wouldn't insist.  Perhaps it could be defined as optional in the base, then `required: true` added to the descriptions for drivers that expect it to come from a devicetree property.  Something like `ngpios` that was recently added in `gpio-controller` (optional in base, required in `arm,mps2-fpgaio-gpio` until all the other controllers are updated).

henrikbrixandersen replied:
@pabigot Good idea. I have changed it to match your suggestion.

pabigot reviewed (approved):
Looks good with the caveats that I'll need to get hardware to test it, and I didn't closely inspect the shell code.  Couple minor comments related to bindings.

vanwinkeljan reviewed (approved):
+1 for the native_posix changes

henrikbrixandersen commented:
I have added a unified driver to AT24 I2C and AT25 SPI EEPROMs. Please re-review as needed.

henrikbrixandersen commented:
From a quick grep though the master source tree, the following supported boards already contain an EEPROM of some sorts:
- 96b_avenger96
- b_l072z_lrwan1
- sam_e70_xplained
- stm3210c_eval
- stm32373c_eval
- stm32f072b_disco
- stm32l476g_disco
- udoo_neo_full_m4

Unfortunately, I do not have access to any of these boards, but if somebody else is looking to test the EEPROM subsystem and driver...

tyrantlizard77 reviewed (commented):
the new subsystem API, syscalls, shell integrayion, etc all look good to me.

But we're missing tests of this subsystem, just adding a case to build_all doesn't give use any code coverage. It especially makes sense to create tests since we can actually exercise it in CI because you provided a native_posix implementation. Could you please implement some test code under tests/drivers/eeprom/ which exercises all these new subsystem APIs, from user mode for the system calls?

tyrantlizard77 reviewed (commented):
the new subsystem API, syscalls, shell integrayion, etc all look good to me.

But we're missing tests of this subsystem, just adding a case to build_all doesn't give use any code coverage. It especially makes sense to create tests since we can actually exercise it in CI because you provided a native_posix implementation. Could you please implement some test code under tests/drivers/eeprom/ which exercises all these new subsystem APIs, from user mode for the system calls? the tests should exercise the full set of return values supported by each API.

henrikbrixandersen committed:
tests: drivers: eeprom: add tests for the EEPROM API

Add EEPROM API test suite.

The test suite is written to run in userspace but it is currently only
whitelisted for native_posix and native_posix_64 boards which do not
support userspace.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
@andrewboie Good point, thank you. I have added a test suite. It run to completion when running without userspace, but fails under userspace. Still trying to figure out what I am doing wrong...
